### PR TITLE
Contact Form: Improve Email Address Handling Personal Data Eraser

### LINKED ISSUE #10076: Contact Form: Erasing feedback using Erase Personal Data does not remove all feedback
Attempting to erase all feedback using `Erase Personal Data` on Jetpack Beta does not remove any feedback.

#### Steps to reproduce the issue

1. Create a contact form on your site
2. Manually edit `modules/contact-form/grunion-contact-form.php` on line 846 to read `$per_page = 1;`
Test:
3. Submit 3 feedbacks through the form, all with the same email address.
4. Go to `wp-admin/ -> Tools -> Erase Personal Data`.
5. Enter the email address you used, click `Send Request`.
6. Find that Pending Request in the table on that page, hover over it and click `Force Erase Personal Data`.
7. See the AJAX requests go, and the success message `All of the personal data found for this user was erased.` appear.
8. Go to `wp-admin/ -> Feedback`
9. Expect to see no feedbacks left with the email you used before.

#### What I expected

All 3 feedbacks to be removed.

#### What happened instead

All 3 feedbacks continue to exist.

### PR DESCRIPTION
**Pull request by mdawaffe**
Fixes #10076 

#### Changes proposed in this Pull Request:

Instead of doing a "normal" WP_Query search, we refine the search with a more complicated query. To do that, we extract the email address from the normal query, and use it to build the more complicated query.

We were not taking into account the fact that string representing the email address can change when used in the query: in particular, because we LIKE (and SQL) escapee the data, the strings representing email addresses with underscores in them change.

When we build the more complicated query, we do another round of LIKE (and SQL) escaping, which is one round too many :)

For an address like `under_score@example.com`, we end up actually searching for `under\\_score`.

Instead of relying on passing the email address around in the query itself, let's just hold on to it in an object property.

Also: some tests :)

#### Testing instructions:

1.`phpunit --filter=WP_Test_Grunion_Contact_Form`

Also:

2. Add a contact form to your site
3. Submit a feedback or two with a specific email address.
4. Go to wp-admin/ → Tools → Personal Data Eraser
5. Enter that email address
6. Hover over that email address's row in the table.
7. Click "Force Erase Personal Data"
8. See the success message
9. See the Feedbacks are gone in wp-admin/ → Feedback

#### Proposed changelog entry for your changes:

Fix bug affecting Contact Form Feedback Personal Data Erasure tool for email addresses containing underscores.

### CODE REVIEW

**Comment by mdawaffe** [comment]
There are still a couple oddities here.

We're searching for `\nAuthor Email: {address}\n` and `\rAuthor Email: {address}\r`, and I'm not sure why.

I believe it's the case that we only need to look for `\n`: https://github.com/Automattic/jetpack/blob/dc3e81bd139f7aeaac80513bf48e2177467f9b0a/modules/contact-form/grunion-contact-form.php#L2350

If there's some line ending oddities I'm not aware of, the two cases we're checking don't cover the `\r\n` case, which would require us to also look for `\nAuthor Email: {address}\r`.

Also, we may still be matching too much with this query. In theory, we could send each about-to-be-deleted feedback through `::parse_fields_from_content()` first to make sure it's a match for the email we're processing. Not sure how performant that would be.

I don't think either issue blocks this PR, but if anyone has thoughts, please voice them :)

**Comment by mdawaffe** [comment]
Fun fact: before I force pushed it away, this branch had a fun side affect. In addition to deleting all the feedbacks from an email address even if that address contained an underscore, it also deleted *all other* feedbacks at the same time. Efficient!

That side affect is gone in commits here on GitHub now :)

File: modules/contact-form/grunion-contact-form.php [hunk 1]
```diff
@@ -955,20 +1010,20 @@ public function personal_data_search_filter( $search ) {
...
-
-			$esc_like_email = esc_sql( $wpdb->esc_like( 'AUTHOR EMAIL: ' . $m[1] ) );
-			$search         = " AND (
-				{$wpdb->posts}.post_content LIKE CONCAT('%', CHAR(13), '{$esc_like_email}', CHAR(13), '%')
-				OR {$wpdb->posts}.post_content LIKE CONCAT('%', CHAR(10), '{$esc_like_email}', CHAR(10), '%')
-			)";
-
-			if ( $last_post_id ) {
-				$search .= $wpdb->prepare( " AND {$wpdb->posts}.ID > %d", $last_post_id );
+		if ( $this->pde_email_address && false !== strpos( $search, '..PDE..AUTHOR EMAIL:..PDE..' ) ) {
```
Comment (mdawaffe) [inline modules/contact-form/grunion-contact-form.php]: We probably don't need the `strpos()` check here. Just check that it's equal to the `..PDE..AUTHOR EMAIL:..PDE..` flag.

**Comment by oskosk** [comment]
Ported to `branch-6.5` in d726ff7fbc2f7c9e87401d7b04dc3e6dce698200

**Pull request merged**